octobot_trading/data/book.py

```diff
@@ -13,42 +13,161 @@
+
+        # Update buy side orders
+        if order[ECOBIC.SIDE.value] == TradeOrderSide.BUY.value:
+            bids = self.get_bids(price)
+            order_index = _order_id_index(order[ECOBIC.ORDER_ID.value], bids)
+            if bids is None or order_index == ORDER_ID_NOT_FOUND:
+                return
+            if new_size != INVALID_PARSED_VALUE:
+                bids[order_index][ECOBIC.SIZE.value] = new_size
+            if new_price != INVALID_PARSED_VALUE:
+                bids[order_index][ECOBIC.PRICE.value] = new_price
+            self._set_bids(price, bids)
+            return
+
+        # Update sell side orders
+        asks = self.get_asks(price)
+        order_index = _order_id_index(order[ECOBIC.ORDER_ID.value], asks)
+        if asks is None or order_index == ORDER_ID_NOT_FOUND:
+            return
+        if new_size != INVALID_PARSED_VALUE:
+            asks[order_index][ECOBIC.SIZE.value] = new_size
+        if new_price != INVALID_PARSED_VALUE:
+            asks[order_index][ECOBIC.PRICE.value] = new_price
+        self._set_asks(price, asks)
+
+    def _set_asks(self, price, asks):
+        self.asks[price] = asks
+
+    def _set_bids(self, price, bids):
+        self.bids[price] = bids
+
+    def _remove_asks(self, price):
+        del self.asks[price]
+
+    def _remove_bids(self, price):
+        del self.bids[price]
+
+    def get_ask(self):
+        return self.asks.peekitem(0)
 
-    def handle_book_delta_delete(self, orders, id_key="id"):
-        ids = [order[id_key] for order in orders]
-        self.orders.drop(index=ids, errors='ignore')
+    def get_bid(self):
+        return self.bids.peekitem(-1)
 
-    def handle_book_delta_update(self, orders, id_key="id"):
-        update_list = pd.json_normalize(orders).set_index(id_key)
-        self.orders.update(update_list)
-        self.orders = self.orders.sort_index(ascending=False)
-        self.timestamp = time()
+    def get_asks(self, price):
+        return self.asks.get(price, None)
 
-    def handle_book_delta_insert(self, orders, id_key="id"):
-        insert_list = pd.json_normalize(orders).set_index(id_key)
-        self.orders.update(insert_list)
-        self.orders = self.orders.sort_index(ascending=False)
+    def get_bids(self, price):
+        return self.bids.get(price, None)
 
-    def get_asks(self, side=TradeOrderSide.SELL.value):
-        return self.orders.query(f"side.str.contains('{side}')", engine='python').values.tolist()
 
-    def get_bids(self, side=TradeOrderSide.BUY.value):
-        return self.orders.query(f"side.str.contains('{side}')", engine='python').values.tolist()
+def _order_id_index(order_id, order_list):
+    """
+    Return order id index in order list
+    :param order_id: the order id to search
+    :param order_list: the order list to check
+    :return: the order id index in order list if found else ORDER_ID_NOT_FOUND
+    """
+    for i in range(order_list):
```

Comment:
Not sure this works, I guess it's one of the reasons for WIP :)

Reply:
Yes :D